lazy-pages/src/common.rs

```diff
@@ -0,0 +1,228 @@
+    pub globals_config: Option<GlobalsConfig>,
+    /// Lazy-pages status: indicates in which mod lazy-pages works actually.
+    pub status: Option<Status>,
+    /// Lazy-pages accesses weights.
+    pub lazy_pages_weights: LazyPagesWeights,
+    /// Cache information about whether page has data in storage
+    pub page_has_data_in_storage: BTreeMap<GranularityPage, bool>,
+}
+
+impl LazyPagesExecutionContext {
+    pub fn is_read_charged(&self, page: GranularityPage) -> bool {
+        self.read_charged.contains(&page)
+    }
+    pub fn is_write_charged(&self, page: GranularityPage) -> bool {
+        self.write_charged.contains(&page)
+    }
+    pub fn set_read_charged(&mut self, page: GranularityPage) -> bool {
+        if self.is_write_charged(page) {
+            false
+        } else {
+            self.read_charged.insert(page)
+        }
+    }
+    pub fn set_write_charged(&mut self, page: GranularityPage) -> bool {
+        self.write_charged.insert(page)
+    }
+    pub fn set_load_data_charged(&mut self, page: GranularityPage) -> bool {
+        self.load_data_charged.insert(page)
+    }
+    pub fn add_to_released(&mut self, page: LazyPage) -> Result<(), Error> {
+        log::trace!("add {page:?} to released");
+        if !self.released_pages.insert(page) {
+            Err(Error::DoubleRelease(page))
+        } else {
+            Ok(())
+        }
+    }
+    pub fn set_program_prefix(&mut self, prefix: Vec<u8>) {
+        self.program_storage_prefix = Some(PagePrefix::new_from_program_prefix(prefix));
+    }
+    pub fn get_key_for_page(&mut self, page: GearPage) -> Result<&[u8], Error> {
+        self.program_storage_prefix
+            .as_mut()
+            .map(|prefix| prefix.calc_key_for_page(page))
+            .ok_or(Error::ProgramPrefixIsNotSet)
+    }
+    pub fn page_has_data_in_storage(&mut self, page: GearPage) -> Result<bool, Error> {
+        if let Some(&res) = self.page_has_data_in_storage.get(&page.to_page()) {
+            return Ok(res);
+        }
+        if sp_io::storage::exists(self.get_key_for_page(page)?) {
+            self.page_has_data_in_storage.insert(page.to_page(), true);
+            Ok(true)
+        } else {
+            self.page_has_data_in_storage.insert(page.to_page(), false);
+            Ok(false)
+        }
+    }
+    pub fn load_page_data_from_storage(
+        &mut self,
+        page: GearPage,
+        buffer: &mut [u8],
+    ) -> Result<(), Error> {
+        if let Some(size) = sp_io::storage::read(self.get_key_for_page(page)?, buffer, 0) {
+            self.page_has_data_in_storage.insert(page.to_page(), true);
+            if size != GearPage::size() {
+                return Err(Error::InvalidPageDataSize {
+                    expected: GearPage::size(),
+                    actual: size,
+                });
+            }
+        } else {
+            self.page_has_data_in_storage.insert(page.to_page(), false);
+        }
+        Ok(())
```

Comment:
I would suggest the following version:
```suggestion
        let Some(size) = sp_io::storage::read(self.get_key_for_page(page)?, buffer, 0) else {
            self.page_has_data_in_storage.insert(page.to_page(), false);
            return Ok(());
        };

        self.page_has_data_in_storage.insert(page.to_page(), true);
        if size != GearPage::size() {
            return Err(Error::InvalidPageDataSize {
                expected: GearPage::size(),
                actual: size,
            });
        }

        Ok(())
```

Reply:
I think this is a less understandable code. `If let Some() else` is bad construction   by my opinion, because `else` is in the end and you can simply miss it when read code